File: deseq/deseq.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

struct DESequence {
  int extra_sp_f;
  int expansion_f;
  int n;
  int size;
  int *arr;
  int start;
  int end;
};

void build(struct DESequence *deseq, int n) {
  assert(n > 0);
  (*deseq).extra_sp_f = 3;
  (*deseq).expansion_f = 3;
  (*deseq).arr = malloc(sizeof(int) * (*deseq).extra_sp_f);
  (*deseq).n = n;
  (*deseq).size = n * (*deseq).extra_sp_f;
  (*deseq).start = n;
  (*deseq).end = 2 * n - 1;
  return;
}

int get_at(struct DESequence *deseq, int idx) {
  assert(idx <= (*deseq).n - 1);
  assert(idx >= 0);
  return (*deseq).arr[(*deseq).start + idx];
}

void set_at(struct DESequence *deseq, int idx, int item) {
  assert(idx <= (*deseq).n - 1);
  assert(idx >= 0);
  (*deseq).arr[(*deseq).start + idx] = item;

  return;
}

void insertl(struct DESequence *deseq, int item) {
  (*deseq).n++;
  if ((*deseq).end < (*deseq).size - 1) {
    (*deseq).end++;
    (*deseq).arr[(*deseq).end] = item;
    return;
  }
  int *new_arr = malloc(sizeof(int) * (*deseq).size * (*deseq).expansion_f);
  for (int i = (*deseq).start; i <= (*deseq).end; i++) {
    new_arr[(*deseq).size + i] = (*deseq).arr[i];
  }
  (*deseq).arr = new_arr;
  (*deseq).start += (*deseq).size;
  (*deseq).end = 2 * (*deseq).size - 1;
  (*deseq).size *= (*deseq).expansion_f;
  (*deseq).end++;
  (*deseq).arr[(*deseq).end] = item;
  return;
}

void insertf(struct DESequence *deseq, int item) {
  (*deseq).n++;
  if ((*deseq).start > 0) {
    (*deseq).start--;
    (*deseq).arr[(*deseq).start] = item;
    return;
  }
  int *new_arr = malloc(sizeof(int) * (*deseq).size * (*deseq).expansion_f);
  for (int i = (*deseq).start; i <= (*deseq).end; i++) {
    new_arr[(*deseq).size + i] = (*deseq).arr[i];
  }
  (*deseq).arr = new_arr;
  (*deseq).start += (*deseq).size;
  (*deseq).end += (*deseq).size;
  (*deseq).size *= (*deseq).expansion_f;
  (*deseq).start--;
  (*deseq).arr[(*deseq).start] = item;
  return;
}

int deletel(struct DESequence *deseq) {
  assert((*deseq).n > 0);
  int deleted = (*deseq).arr[(*deseq).end];
  (*deseq).n--;
  (*deseq).end--;
  return deleted;
}

int deletef(struct DESequence *deseq) {
  assert((*deseq).n > 0);
  int deleted = (*deseq).arr[(*deseq).start];
  (*deseq).n--;
  (*deseq).start++;
  return deleted;
}

int size(struct DESequence *deseq) { return (*deseq).size; }

int len(struct DESequence *deseq) { return (*deseq).n; }

void print(struct DESequence *deseq) {
  for (int i = (*deseq).start; i <= (*deseq).end; i++) {
    printf("%d ", (*deseq).arr[i]);
  }
  printf("\n");
  return;
}

void dump(struct DESequence *deseq) {
  for (int i = 0; i < (*deseq).size; i++) {
    printf("%d ", (*deseq).arr[i]);
  }
  printf("\n");
  return;
}

int main() {
  int n = 1;
  struct DESequence deseq;
  build(&deseq, n);
  set_at(&deseq, 0, -1);
  printf("\n");
  for (int i = 1; i < 10; i++) {
    insertf(&deseq, i);
  }
  for (int i = 1; i < 10; i++) {
    deletel(&deseq);
    printf("\n%d\n", len(&deseq));
    printf("\n%d\n", size(&deseq));
    dump(&deseq);
    print(&deseq);
  }
  return 0;
}

File: deseq/deseq.py
# split data in half, reverse first half and
# store in a list, and store second half in a list

class DESequence:
    def __init__(self):
        self.n = 0
        self.__lft = []
        self.__rgt = []

    def build(self, lst, n):
        if len(lst) == 0:
            self.__lft.reverse()
            return self
        self.n += 1
        if self.n <= n // 2:
            self.__lft.append(lst[0])
        else:
            self.__rgt.append(lst[0])
        return self.build(lst[1:], n)

    def get_at(self, idx):
        return self.__lft[len(self.__lft) - 1 - idx] if (
            idx < len(self.__lft)) else self.__rgt[
            idx - len(self.__lft)]

    def insertl(self, item):
        self.n += 1
        self.__rgt.append(item)

    def insertf(self, item):
        self.n += 1
        self.__lft.append(item)

    def deletel(self):
        if len(self.__rgt) <= 0:
            self.rebuild([], self.n)
        self.n -= 1
        self.__rgt.pop()

    def deletef(self):
        if len(self.__lft) <= 0:
            self.__rgt.reverse()
            self.rebuild([], self.n)
        self.n -= 1
        return self.__lft.pop()

    def rebuild(self, lst, n):
        if self.n <= 0:
            self.build(lst, len(lst))
            return
        self.n -= 1
        if len(self.__lft) > 0:
            lst.append(self.__lft[-1])
            self.__lft.pop()
        else:
            lst.append(self.__rgt[-1])
            self.__rgt.pop()
        self.rebuild(lst, n)

def print_all(d):
    for i in range(d.n):
        print(d.get_at(i), end=" ")

def main():
    lst = [1, 2, 3, 4, 5, 6]
    d = DESequence().build(lst, len(lst))
    print_all(d)
    for i in range(13, 20):
        d.insertl(i)
    print("\n\n")
    print_all(d)

    lst = [1, 2, 3, 4, 5, 6]
    d = DESequence().build(lst, len(lst))
    for i in range(10):
        d.insertf(-i)
    print("\n\n")
    print_all(d)

    for i in range(3):
        d.deletel()
    print("\n\n")
    print_all(d)

    for i in range(10):
        d.deletef()
    print("\n\n")
    print_all(d)

    print("\n")

if __name__ == "__main__":
    main()
